### PR TITLE
oracle: add oracle web3 endpoint pool

### PR DESCRIPTION
**Pull request by jordipainan**
If an Oracle that is subscribed to one or more
ethereum smart contract events
have any problem with the endpoint that is
connected with, after some retries, another
web3 endpoint will be used.

If there is not any web3 endpoint working
the application will terminate.

When a new connection is created the ethevents
service will be restarted as well as the
PrintInfo go routine.

A context.WithCancel is being used to terminate
all the goroutines created by the ethereum events
service and the contracts will be resolved on
the ENS an instanciated again. That way if any
event is lost during the time that there
are connection issues, will be catch again
on restart.

### CODE REVIEW

File: cmd/dvotenode/dvotenode.go [hunk 1]
```diff
@@ -543,6 +543,28 @@ func main() {
...
+					w3uris = append(w3uris, web3EndpointTrimmed)
+				case strings.HasSuffix(web3EndpointTrimmed, "ipc"):
+					w3uris = append(w3uris, web3EndpointTrimmed)
+				default:
+					log.Warnf(`invalid web3 endpoint %s must be websocket or IPC
+						for event subscription`, web3EndpointTrimmed)
+				}
+			}
+		} else {
+			log.Errorf(`No web3 endpoint found, oracle needs at least
```
Comment (p4u) [inline cmd/dvotenode/dvotenode.go]: That should probably be a Fatal... or if its an Error do something

File: cmd/dvotenode/dvotenode.go [hunk 2]
```diff
@@ -543,6 +543,28 @@ func main() {
...
+			one valid web3 endpoint to handle Ethereum contract events`)
+		}
+
+		log.Debugf("Using %+v web3 endpoints", w3uris)
```
Comment (p4u) [inline cmd/dvotenode/dvotenode.go]: capital letter

File: cmd/dvotenode/dvotenode.go [hunk 3]
```diff
@@ -543,6 +543,28 @@ func main() {
 			log.Fatal(err)
 		}
 
+		w3uris := []string{}
+		if len(globalCfg.W3Config.W3External) > 0 {
```
Comment (p4u) [inline cmd/dvotenode/dvotenode.go]: I'd revert that logic... its a bit messy.

Remove the `if len(globalCfg.W3Config.W3External) > 0 {` and also the else and just add after the for

`if len(w3uris) == 0 { log.Fatal("no valid web3 endpoint found") }`

File: cmd/dvotenode/dvotenode.go (hunk 1 above)
Comment (p4u) [reply cmd/dvotenode/dvotenode.go:562]: superseded by the previous comment

**Review by p4u — CHANGES_REQUESTED** [review]

**Comment by p4u** [comment]
@arnaucube and @mvdan can you both review too? thanks!

File: cmd/dvotenode/dvotenode.go [hunk 4]
```diff
@@ -543,6 +543,27 @@ func main() {
...
+				log.Warnf(`invalid web3 endpoint %s must be websocket or IPC
+						for event subscription`, web3EndpointTrimmed)
+			}
+		}
+
+		if len(w3uris) == 0 {
+			log.Fatal(`no web3 endpoint found, oracle needs at least
+			one valid web3 endpoint to handle Ethereum contract events`)
+		}
+		log.Debugf("using %+v web3 endpoints", w3uris)
```
Comment (p4u) [inline cmd/dvotenode/dvotenode.go]: this should be info IMO

File: cmd/dvotenode/dvotenode.go [hunk 5]
```diff
@@ -559,56 +580,45 @@ func main() {
...
-					}
-				}
-				if len(whiteListedAddr) > 0 {
-					log.Infof("ethereum whitelisted addresses %+v", whiteListedAddr)
+			var initBlock *int64
+			if !globalCfg.EthEventConfig.SubscribeOnly {
+				initBlock = new(int64)
+				chainSpecs, err := ethchain.SpecsFor(globalCfg.W3Config.ChainType)
+				if err != nil {
+					log.Warn("cannot get chain block to start looking for events, using 0")
```
Comment (p4u) [inline cmd/dvotenode/dvotenode.go]: I think initBlock should not be a pointer, what's the reason? 

Also I'm not sure we should uset initBlock=0 if error. I think its safer to do a Fatal() and print the error instead

File: cmd/dvotenode/dvotenode.go [hunk 6]
```diff
@@ -559,56 +580,45 @@ func main() {
...
+					*initBlock = 0
+				} else {
```
Comment (p4u) [inline cmd/dvotenode/dvotenode.go]: No need for this else if err executes Fatal()

File: cmd/dvotenode/dvotenode.go [hunk 7]
```diff
@@ -559,56 +580,45 @@ func main() {
...
-					vnode,
-					evh,
-					sc,
-					whiteListedAddr); err != nil {
-					log.Fatal(err)
+			}
+
+			whiteListedAddr := []string{}
+			for _, addr := range globalCfg.VochainConfig.EthereumWhiteListAddrs {
+				if ethcommon.IsHexAddress(addr) {
```
Comment (p4u) [inline cmd/dvotenode/dvotenode.go]: If `addr` is not Hex, shouldn't we at least print a Warning?

File: cmd/dvotenode/dvotenode.go [hunk 8]
```diff
@@ -632,7 +642,7 @@ func main() {
 				log.Fatal(err)
 			}
 			if err := apior.EnableERC20(globalCfg.W3Config.ChainType,
```
Comment (p4u) [inline cmd/dvotenode/dvotenode.go]: I think breaking lines here is not needed

File: ethereum/ethevents/events.go [hunk 9]
```diff
@@ -63,6 +63,7 @@ type EthereumEvents struct {
 	ContractsAddress []common.Address
 	// ContractsInfo holds useful info for working with the desired contracts
 	ContractsInfo map[string]*ethereumhandler.EthereumContract
+	StartBlock    *int64
```
Comment (p4u) [inline ethereum/ethevents/events.go]: why a pointer?

Comment (p4u) [reply ethereum/ethevents/events.go]: If you thinking about saving memory, take into account you are actually doing the opposite sice a pointer will require 2xint64 (one for the address and another for the value)

File: ethereum/ethevents/events.go [hunk 10]
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
 		go ev.runEventProcessor(ctx)
 	}
 
+	// check eth connection
+	connCh := make(chan bool, 1)
+	go func(chan bool) {
+		for {
+			time.Sleep(time.Second * 5)
+			blockTctx, cancel := context.WithTimeout(ctx, types.EthereumReadTimeout*2)
```
Comment (p4u) [inline ethereum/ethevents/events.go]: Why EthereumReadTimeout x2 ?

Comment (p4u) [inline ethereum/ethevents/events.go]: Just call it ctx (more standard)

File: ethereum/ethevents/events.go [hunk 11]
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
...
+			if _, err := ev.VotingHandle.EthereumClient.BlockByNumber(blockTctx, nil); err != nil {
+				log.Warnf("cannot check ethereum connection while running the event processor, %s", err)
+				connCh <- false
+				cancel()
```
Comment (p4u) [inline ethereum/ethevents/events.go]: cancel() should also be called if not error

File: ethereum/ethevents/events.go [hunk 12]
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
 		go ev.runEventProcessor(ctx)
 	}
 
+	// check eth connection
+	connCh := make(chan bool, 1)
+	go func(chan bool) {
+		for {
+			time.Sleep(time.Second * 5)
+			blockTctx, cancel := context.WithTimeout(ctx, types.EthereumReadTimeout*2)
+			if _, err := ev.VotingHandle.EthereumClient.BlockByNumber(blockTctx, nil); err != nil {
```
Comment (p4u) [inline ethereum/ethevents/events.go]: Shouldn't the block number be saved and check that it is actually being increased? Else we won't detect a frozen web3 endpoint

File: ethereum/ethevents/events.go [hunk 13]
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
 		go ev.runEventProcessor(ctx)
 	}
 
+	// check eth connection
+	connCh := make(chan bool, 1)
```
Comment (p4u) [inline ethereum/ethevents/events.go]: This channel should have a more descriptive name, such as "connFailure"

File: ethereum/ethevents/events.go [hunk 14]
```diff
@@ -377,20 +400,28 @@ func (ep *EventProcessor) next() *ethtypes.Log {
...
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				if event.Removed {
+					log.Warnf("removing reversed log event: %s", evtJSON)
+					ev.EventProcessor.del(&event)
+				} else {
+					if event.BlockNumber == 0 {
+						log.Fatalf("this event %+v must not be arribe here (BOMB)", event)
```
Comment (p4u) [inline ethereum/ethevents/events.go]: what's the meaning of this message? I don't get it. Can you rephrase?

File: ethereum/handler/ethereumHandler.go [hunk 15]
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
...
-			maxtries--
-			continue
+	for _, endpoint := range dialEndpoints {
+		maxtries := 5
+		for {
+			if maxtries == 0 {
+				log.Warnf("could not connect to %s endpoint, trying the next one ...", endpoint)
+				break
+			}
+			eh.EthereumRPC, err = ethrpc.Dial(endpoint)
```
Comment (p4u) [inline ethereum/handler/ethereumHandler.go]: Use inline if

File: ethereum/handler/ethereumHandler.go [hunk 16]
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
...
-		if err != nil || eh.EthereumRPC == nil {
-			log.Warnf("cannot create an ethereum rpc connection: (%s), trying again ...", err)
-			time.Sleep(time.Second * 2)
-			maxtries--
-			continue
+	for _, endpoint := range dialEndpoints {
+		maxtries := 5
+		for {
+			if maxtries == 0 {
+				log.Warnf("could not connect to %s endpoint, trying the next one ...", endpoint)
```
Comment (p4u) [inline ethereum/handler/ethereumHandler.go]: "one..." (or even better, remove the "..." since we are not using that in general)

File: ethereum/handler/ethereumHandler.go [hunk 17]
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
...
+				break
+			}
+			eh.EthereumRPC, err = ethrpc.Dial(endpoint)
+			if err != nil || eh.EthereumRPC == nil {
+				log.Warnf("cannot create an ethereum rpc connection with : (%s), trying again ...", err)
```
Comment (p4u) [inline ethereum/handler/ethereumHandler.go]: `"with %s: (%v)", endpoint, err)`

File: ethereum/handler/ethereumHandler.go [hunk 18]
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
...
+				time.Sleep(time.Second * 5)
```
Comment (p4u) [inline ethereum/handler/ethereumHandler.go]: Why sleeping 5 seconds? Seems too much

File: ethereum/handler/ethereumHandler.go [hunk 19]
```diff
@@ -163,7 +166,10 @@ func (eh *EthereumHandler) PrintInfo(ctx context.Context, seconds time.Duration)
 		info, err = eh.SyncInfo(tctx)
 		cancel()
 		if err != nil {
-			log.Warn(err)
+			log.Warnf("error getting ethereum info: %s", err)
+			if err := eh.Connect(eh.Endpoints); err != nil {
+				log.Warnf("cannot connect to any web3 endpoint: %v", err)
```
Comment (p4u) [inline ethereum/handler/ethereumHandler.go]: Shouldn't this Fatal?

File: service/ethevents.go [hunk 20]
```diff
@@ -40,20 +41,37 @@ func EthEvents(
...
 		return fmt.Errorf("couldn't create ethereum events listener: %w", err)
 	}
 	for _, e := range evh {
 		ev.AddEventHandler(e)
 	}
-	go func() {
-		ev.SubscribeEthereumEventLogs(ctx, startBlock)
-	}()
+
+	initialized := new(bool)
```
Comment (p4u) [inline service/ethevents.go]: Why a pointer?

File: service/ethevents.go [hunk 21]
```diff
@@ -40,20 +41,37 @@ func EthEvents(
...
+	go func(initialized *bool) {
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+			ctx, cancel := context.WithCancel(ctx)
+			if *initialized {
+				if ev.VotingHandle, err = ethereumhandler.NewEthereumHandler(ev.ContractsInfo, specs.NetworkSource, w3uris); err != nil {
```
Comment (p4u) [inline service/ethevents.go]: This line is 126 chars

File: service/ethevents.go [hunk 22]
```diff
@@ -40,20 +41,37 @@ func EthEvents(
...
+					log.Fatalf("cannot restart ethereum events, cannot create ethereum handler: %v", err)
+				}
+			}
+			*initialized = true
```
Comment (p4u) [inline service/ethevents.go]: I don't understand the logic behind this routine. Why a for loop? "initialized" is set to true on each iteration, why? Also why ev.VotingHandle is inside the for loop?

**Review by p4u — CHANGES_REQUESTED** [review]

File: ethereum/handler/ethereumHandler.go [hunk 23]
```diff
@@ -84,32 +87,29 @@ func NewEthereumHandler(contracts map[string]*EthereumContract, srcNetworkId mod
...
+			}
+			eh.EthereumRPC, err = ethrpc.Dial(endpoint)
+			if err != nil || eh.EthereumRPC == nil {
+				log.Warnf("cannot create an ethereum rpc connection with : (%s), trying again ...", err)
+				time.Sleep(time.Second * 5)
+				maxtries--
+				continue
+			}
+			// if RPC connection established, create an ethereum client using the RPC client
+			eh.EthereumClient = ethclient.NewClient(eh.EthereumRPC)
```
Comment (mvdan) [inline ethereum/handler/ethereumHandler.go:107]: are you sure that the EthereumClient field isn't racy? it's used from other goroutines as far as I can tell.

File: ethereum/ethevents/events.go [hunk 24]
```diff
@@ -290,13 +290,36 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
...
+		case <-sub.Err():
 			// TODO: @jordipainan do not fatal here, handle error on event subscription channel
-			log.Fatalf("ethereum events subscription error on channel: %s", err)
+			log.Warn("ethereum events subscription error on channel")
+			log.Infof("restarting ethereum events subscription with web3: %+v", ev.DialAddrs)
+			ev.EventProcessor.eventProcessorRunning = false
+			return
 		case event := <-logs:
 			ev.EventProcessor.Events <- event
+		case <-connCh:
```
Comment (mvdan) [inline ethereum/ethevents/events.go]: we have two identical cases in this select now; why do we have two channels instead of one?

File: cmd/dvotenode/dvotenode.go (hunk 4 above)
Comment (jordipainan) [reply cmd/dvotenode/dvotenode.go]: changed to info :)

File: cmd/dvotenode/dvotenode.go (hunk 5 above)
Comment (jordipainan) [reply cmd/dvotenode/dvotenode.go]: yes agree, at some point we discussed to not fatal in this kind of cases, that is why most of this errors are handled with a warning. In any case I also prefer to not use initBlock=0

File: cmd/dvotenode/dvotenode.go (hunk 7 above)
Comment (jordipainan) [reply cmd/dvotenode/dvotenode.go]: yes, added

File: ethereum/ethevents/events.go (hunk 10 above)
Comment (jordipainan) [reply ethereum/ethevents/events.go]: copy paste err, changed

File: ethereum/ethevents/events.go (hunk 11 above)
Comment (jordipainan) [reply ethereum/ethevents/events.go]: moved and added a defer that will be executed in the very end

File: ethereum/ethevents/events.go (hunk 12 above)
Comment (jordipainan) [reply ethereum/ethevents/events.go]: there is an else clause that will be executed when restarting the subscription (100 blocks), I think is enough for this purpose. What do you think?

File: ethereum/ethevents/events.go (hunk 13 above)
Comment (jordipainan) [reply ethereum/ethevents/events.go]: changed :)

File: ethereum/ethevents/events.go (hunk 14 above)
Comment (jordipainan) [reply ethereum/ethevents/events.go]: just a sanity check cause I do not expect having block number 0 on event logs. Changed the message for more clarity

File: ethereum/handler/ethereumHandler.go (hunk 18 above)
Comment (jordipainan) [reply ethereum/handler/ethereumHandler.go]: lowered to 3

File: ethereum/handler/ethereumHandler.go (hunk 19 above)
Comment (jordipainan) [reply ethereum/handler/ethereumHandler.go]: yes assuming last comment made on this, changed to a fatal

File: service/ethevents.go (hunk 20 above)
Comment (jordipainan) [reply service/ethevents.go]: no need, only one goroutine use the var, changed and no longer using a pointer

File: service/ethevents.go (hunk 22 above)
Comment (jordipainan) [reply service/ethevents.go]: kind of a workaround because new EthEvents already invokes new ethereum handler and I wanted to avoid this. Solving properly requires some refactor on the new EthEvents function. This is out of the scope of this commit.
WIll add a TODO explaining more about this for adding clarity and i'll make another PR improving this.

File: ethereum/ethevents/events.go (hunk 24 above)
Comment (jordipainan) [reply ethereum/ethevents/events.go]: discussed privately and solved

Comment (jordipainan) [reply ethereum/ethevents/events.go]: btw, the fallthrough clause cannot be applied here afaik. The compiler complains about it and I only solved it with a weird if clause.

File: ethereum/handler/ethereumHandler.go (hunk 23 above)
Comment (jordipainan) [reply ethereum/handler/ethereumHandler.go:107]: yes it is, adding a TODO for having 2 different connections:
- One for the subscription
- One for the ethereum sync status

File: cmd/dvotenode/dvotenode.go (hunk 5 above)
Comment (p4u) [reply cmd/dvotenode/dvotenode.go]: If an error is something that does not allow the execution, it should be a Fatal. And still, why initBlock is a pointer?

File: cmd/dvotenode/dvotenode.go [hunk 25]
```diff
@@ -545,6 +545,27 @@ func main() {
...
+				w3uris = append(w3uris, web3EndpointTrimmed)
+			default:
+				log.Warnf(`invalid web3 endpoint %s must be websocket or IPC
+						for event subscription`, web3EndpointTrimmed)
+			}
+		}
+
+		if len(w3uris) == 0 {
+			log.Fatal(`no web3 endpoint found, oracle needs at least
+			one valid web3 endpoint to handle Ethereum contract events`)
```
Comment (mvdan) [inline cmd/dvotenode/dvotenode.go]: FYI this log line will have a newline and lots of whitespace. I'd just join it as a single line, but if you really want to split it:

```
log.Fatal("foo bar " + 
    "baz")
```

File: cmd/dvotenode/dvotenode.go [hunk 26]
```diff
@@ -561,56 +582,48 @@ func main() {
...
+				context.Background(),
+				w3uris,
+				globalCfg.W3Config.ChainType,
+				initBlock,
+				cm,
+				signer,
+				vnode,
+				evh,
+				sc,
+				whiteListedAddr); err != nil {
```
Comment (mvdan) [inline cmd/dvotenode/dvotenode.go]: I really think we're going way, way overkill with avoiding long lines. You're literally writing multiple lines which are >80% whitespace :) This entire call can very easily fit in 2-3 lines.

File: ethereum/ethevents/events.go [hunk 27]
```diff
@@ -290,13 +291,38 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
...
+			}
+			ev.EthereumLastKnownBlock = block.Number().Int64()
+		}
+	}(connFailure)
+
 	for {
 		select {
-		case err := <-sub.Err():
+		case <-sub.Err():
 			// TODO: @jordipainan do not fatal here, handle error on event subscription channel
```
Comment (mvdan) [inline ethereum/ethevents/events.go]: I think this TODO is outdated now

File: ethereum/handler/ethereumHandler.go (hunk 23 above)
Comment (mvdan) [reply ethereum/handler/ethereumHandler.go:107]: Hm, OK. Seems like a pretty big TODO, but I guess "it usually works" is better than delaying this PR.

**Review by mvdan — APPROVED** [review]
SGTM, just a few minor nits

File: ethereum/ethevents/events.go (hunk 24 above)
Comment (mvdan) [reply ethereum/ethevents/events.go]: You're right, I forgot that fallthroughs can't be used in selects. I guess not worth doing anything more magic here, but we could at least put both cases next to each other.

**Comment by jordipainan** [comment]
thanks for the re-review @mvdan

File: ethereum/ethevents/events.go (hunk 12 above)
Comment (jordipainan) [reply ethereum/ethevents/events.go]: will tackle on the next PR, added a better approach :)

File: service/ethevents.go (hunk 22 above)
Comment (jordipainan) [reply service/ethevents.go]: will be better addressed on the next pr

File: ethereum/ethevents/events.go [hunk 28]
```diff
@@ -290,13 +291,37 @@ func (ev *EthereumEvents) SubscribeEthereumEventLogs(ctx context.Context, fromBl
...
+			time.Sleep(time.Second * 5)
+			tctx, cancel := context.WithTimeout(ctx, types.EthereumReadTimeout)
+			defer cancel()
+			block, err := ev.VotingHandle.EthereumClient.BlockByNumber(tctx, nil)
+			if err != nil {
+				log.Warnf("cannot check ethereum connection while running the event processor, %s", err)
+				connFailure <- false
+				return
+			}
+			ev.EthereumLastKnownBlock = block.Number().Int64()
```
Comment (p4u) [inline ethereum/ethevents/events.go]: This var is never used and also it looks like racy, I think sync.Atomic should be used here

Comment (jordipainan) [reply ethereum/ethevents/events.go]: yep, so finally adding the commit I mentioned regarding this case here. And yes, is used on the resubscribe routine

**Review by p4u — APPROVED** [review]

File: ethereum/ethevents/events.go [hunk 29]
```diff
@@ -172,155 +161,81 @@ func (ev *EthereumEvents) AddEventHandler(handler EventHandler) {
...
-			time.Sleep(time.Second * 2)
-			if err := ev.VotingHandle.Connect(ev.DialAddrs); err != nil {
-				log.Warn(err)
-			}
-			continue
-		}
-		lastBlock = blk.Number().Int64()
-		break
+	var err error
+	uLastBlockNumber, err := ev.VotingHandle.EthereumClient.BlockNumber(tctx)
```
Comment (mvdan) [inline ethereum/ethevents/events.go:170]: you only need to do this remote ethereum request if fromBlock==0, right? I'd move it inside the "if".

File: ethereum/ethevents/events.go [hunk 30]
```diff
@@ -172,155 +161,81 @@ func (ev *EthereumEvents) AddEventHandler(handler EventHandler) {
...
-		// For security, even if subscribe only, force to process at least the past `readBlockPast`
-		// Same as the last processEventLogsFromTo function call
-		if err := ev.processEventLogsFromTo(ctx, blk.Number().Int64()-readBlocksPast,
-			blk.Number().Int64(), ev.VotingHandle.EthereumClient); err != nil {
-			log.Errorf("cannot process event logs: %v", err)
-		}
+	lastBlockNumber := int64(uLastBlockNumber)
+	fromBlock := ev.EthereumLastKnownBlock.Load().(int64)
+	if fromBlock == 0 {
+		ev.EthereumLastKnownBlock.Store(lastBlockNumber)
```
Comment (mvdan) [inline ethereum/ethevents/events.go:177]: note that this Store is separate from the Load, so it's not atomic. I imagine we don't mind that. If the original atomic value was 0 and two goroutines both try to update it at the same time, I imagine they'd both try to update it to the same new value. If it's possible that those two goroutines would end up with different "latest block" values, then that might be problematic.

File: ethereum/handler/ethereumHandler.go [hunk 31]
```diff
@@ -161,15 +161,15 @@ func (eh *EthereumHandler) PrintInfo(ctx context.Context, seconds time.Duration)
 	var err error
 	var syncingInfo string
 	for {
+		if ctx.Err() != nil {
+			return
+		}
 		time.Sleep(seconds)
```
Comment (mvdan) [inline ethereum/handler/ethereumHandler.go:167]: nit: do a select between ctx.Done and time.After, to stop sleeping if the context is cancelled

**Review by mvdan — COMMENTED** [review]
Sorry for the late review. Better late than never :) Mostly just nits.

**Comment by p4u** [comment]
@mvdan that's old code, there is another commit that modifies the behavior of this PR. I'd propose (if you want) to review the code in master and create a new PR if you have suggestions.

**Comment by mvdan** [comment]
Yep I know - Jordi had asked me to review the last commit two weeks ago, but I was away for some time, so I'm only getting to it now :)
